kernel/src/table_changes/scan.rs

```diff
@@ -0,0 +1,203 @@
+use std::sync::Arc;
+
+use itertools::Itertools;
+use tracing::debug;
+
+use crate::scan::ColumnType;
+use crate::schema::SchemaRef;
+use crate::{DeltaResult, ExpressionRef};
+
+use super::{TableChanges, CDF_FIELDS};
+
+/// The result of building a [`TableChanges`] scan over a table. This can be used to get a change
+/// data feed from the table
+#[allow(unused)]
+pub struct TableChangesScan {
+    table_changes: Arc<TableChanges>,
+    logical_schema: SchemaRef,
+    predicate: Option<ExpressionRef>,
+    all_fields: Vec<ColumnType>,
+    have_partition_cols: bool,
+}
+
+/// Builder to read the `TableChanges` of a table.
+pub struct TableChangesScanBuilder {
+    table_changes: Arc<TableChanges>,
+    schema: Option<SchemaRef>,
+    predicate: Option<ExpressionRef>,
+}
+
+impl TableChangesScanBuilder {
+    /// Create a new [`TableChangesScanBuilder`] instance.
+    pub fn new(table_changes: impl Into<Arc<TableChanges>>) -> Self {
+        Self {
+            table_changes: table_changes.into(),
+            schema: None,
+            predicate: None,
+        }
+    }
+
+    /// Provide [`Schema`] for columns to select from the [`TableChanges`].
+    ///
+    /// A table with columns `[a, b, c]` could have a scan which reads only the first
+    /// two columns by using the schema `[a, b]`.
+    ///
+    /// [`Schema`]: crate::schema::Schema
+    pub fn with_schema(mut self, schema: SchemaRef) -> Self {
+        self.schema = Some(schema);
+        self
+    }
+
+    /// Optionally provide a [`SchemaRef`] for columns to select from the [`TableChanges`]. See
+    /// [`TableChangesScanBuilder::with_schema`] for details. If `schema_opt` is `None` this is a no-op.
+    pub fn with_schema_opt(self, schema_opt: Option<SchemaRef>) -> Self {
+        match schema_opt {
+            Some(schema) => self.with_schema(schema),
+            None => self,
+        }
```

Comment:
does this also work?
```suggestion
        schema_opt.map_or(self, |schema| self.with_schema(schema))
```

Reply:
Ah it doesn't because it sees self is moving ownership to `map_or`'s default and also to the closure. Not allowed 😔

Reply:
`map_or_else` might delay the move enough for it to work?
```rust
schema_opt.map_or_else(|| self, |schema| self.with_schema(schema))
```